## Contents

* [mixer_uni_table.sh](#mixer_uni_tablesh)
* [mixer_bi_table.sh](#mixer_bi_tablesh)
* [fdr_fuma_snp_input.sh](#fdr_fuma_snp_inputsh)
* [fdr_fuma_snp_table.sh](#fdr_fuma_snp_tablesh)
* [fdr_fuma_loci_table.sh](#fdr_fuma_loci_tablesh)
* [fdr_fuma_gene_table.sh](#fuma_gene_tablesh)
* [fdr_fuma_go_table.sh](#fuma_go_tablesh)
* [fdr_fuma_eqtl_table.sh](#fdr_fuma_eqtl_tablesh)
* [fdr_fuma_trans_loci_table.sh](#fdr_fuma_trans_loci_tablesh)

## Summary of scripts

### mixer_uni_table.sh

**Function**
This script extracts data from mixer analysis for supplementary univariate
table.

**Usage** ``sh mixer_uni_table.sh uni_csv outfile``

**Arguments**
* `uni_csv` - csv file containing univariate info
* `outfile` - output file

### mixer_bi_table.sh

**Function**
This script extracts data from mixer analysis for supplementary bivariate
table.

**Usage** ``sh mixer_bi_table.sh bi_csv outfile``

**Arguments**
* `bi_csv` - csv file containing bivariate info
* `outfile` - output file

### fdr_fuma_snp_input.sh

**Function**
This script generates a snp list from PleioFDR output for FUMA analysis.

**Usage** ``sh fdr_fuma_snp_input.sh fdr_clump_snp_file fdr R2 outfile``

**Arguments**
* `fdr_clump_snp_file` - fdr clumping snp file
* `fdr` - FDR filter for selecting snps
* `R2` - PleioFDR R2 filter for selecting snps
* `outfile` - output file including selected snp list

**Example**
```
sh fdr_fuma_snp_input.sh conj.result.clump.snps.csv 0.1 0.6 TRAIT1_vs_TRAIT2_input_snps.txt
```

### fdr_fuma_snp_table.sh

**Function**
This script generates data for supplementary snp table (FUMA+SUMSTAT)
based on selected snp subset (fdr,r2) from PleioFDR/FUMA analysis.
NOTE: there might be difference between the loci numbers from PleioFDR analysis
and the ones from FUMA analysis (GenomicLocus). This is because the former
is based on clumping of PleioFDR, while the latter is based on clumping of
p-value, good to distiguish between them.

**Usage** ``sh fdr_fuma_snp_table.sh fdr_clump_snp_file fdr r2 fuma_snp_file sumstat1 sumstat2 tag1 tag2 outfolder``

**Arguments**
* `fdr_clump_snp_file` - fdr clumping snp file
* `fdr` - PleioFDR filter for selecting snps
* `r2` - r2 filter for selecting snps
* `fuma_snp_file` - fuma annotated snp file
* `sumstat1` - primary summary statistic file
* `sumstat2` - secondary summary statistic file
* `tag1` - tag of trait1
* `tag2` - tag of trait2
* `outfolder` - output folder

**Example**
```
sh fdr_fuma_snp_table.sh conj.result.clump.snps.csv 0.1 0.6 snps.txt sumstat/std/trait1.sumstats.gz sumstat/std/trait2.sumstats.gz tag1 tag2 outfolder
```

### fdr_fuma_loci_table.sh

**Function**
This script generates data for supplementary loci table (PleioFDR+FUMA+SUMSTAT)
from PleioFDR/FUMA analysis.

**Usage** ``sh fdr_fuma_loci_table.sh fdr_clump_loci_file fdr_fuma_snp_table outfile [trait1locifile trait2locifile] [fuma_gwasc keywords1 keywords2]``

**Arguments**
* `fdr_clump_loci_file` - fdr clumping loci file
* `fdr_fuma_snp_table` - TRAIT1_vs_TRAIT2_snps.txt file generated by script fdr_fuma_snp_table.sh
* `outfile` - output file, recommanded to name as TRAIT1_vs_TRAIT2_loci.txt
* `trait1locifile` - old loci file with 3 or 4 columns (CHR, MinBP, MaxBP, [leadSNP]) of trait1, use '-' for none
* `trait2locifile` - old loci file with 3 or 4 columns (CHR, MinBP, MaxBP, [leadSNP]) of trait2, use '-' for none
* `fuma_gwasc` - fuma gwascatalog file, use '-' for none
* `keywords1` - quoted keywords of trait1 combined with &, delimited by |, use '-' for none
* `keywords2` - quoted keywords of trait2 combined with &, delimited by |, use '-' for none

**Example**
```
sh fdr_fuma_loci_table.sh conj.result.clump.loci.csv ADHD_vs_MOOD_snps.txt ADHD_vs_MOOD_loci.txt adhd_gwas_loci.csv mood_gwas_loci.csv gwascatalog.txt 'attention & deficit & hyperact' 'mood & instab'
```

### fdr_fuma_gene_table.sh

**Function**
This script generates data for supplementary gene table from PleioFDR/FUMA analysis.

**Usage** ``sh fuma_gene_table.sh fdr_fuma_snp_table fuma_gene_file outfile``

**Arguments**
* `fdr_fuma_snp_table` - TRAIT1_vs_TRAIT2_snps.txt file generated by script fdr_fuma_snp_table.sh
* `fuma_gene_file` - file that contains fuma gene info
* `outfile` - output file

**Example**
```
sh fdr_fuma_gene_table.sh TRAIT1_vs_TRAIT2_snps.txt genes.txt TRAIT1_vs_TRAIT2_genes.txt
```

### fdr_fuma_go_table.sh

**Function**
This script generates data for supplementary gene ontology gene-sets table from FUMA analysis.

**Usage** ``sh fdr_fuma_go_table.sh fuma_gs_file outfile``

**Arguments**
* `fuma_gs_file` - file that contains fuma gene-sets info
* `outfile` - output file

**Example**
```
sh fuma_go_table.sh GS.txt TRAIT1_vs_TRAIT2_genes.txt
```

### fdr_fuma_eqtl_table.sh

**Function**
This script generates data for supplementary eqtl table from PleioFDR/FUMA analysis.

**Usage** ``sh fdr_fuma_eqtl_table.sh fdr_clump_loci_file fuma_eqtl_file outfile``

**Arguments**
* `fdr_clump_loci_file` - fdr clumping loci file
* `fuma_eqtl_table` - file that contains fuma eqtl info
* `outfile` - output file

**Example**
```
sh fuma_eqtl_table.sh conj.result.clump.loci.csv eqtl.txt TRAIT1_vs_TRAIT2_eqtl.txt
```

### fdr_fuma_trans_loci_table.sh

**Function**
This script identifies transdiagnostic loci shared across multiple traits from PleioFDR/FUMA analysis.

**Usage** ``sh fdr_fuma_trans_loci.sh list_of_loci_files outfile``

**Arguments**
* `list_of_loci_files` - file that includes paths to shared loci file, and fdr_fuma_snp_table, make sure that absolute or relative path to be specified
* `outfile` - out file that contains transdiagnostic loci fdr/fuma results

**Example**
```
sh fdr_fuma_trans_loci.sh mood_psych_shared_loci_conj005.txt mood_psych_conj005_trans_loci.txt

mood_psych_shared_loci_conj005.txt: https://github.com/precimed/yunhoop/blob/master/config/mood_psych_shared_loci_conj005.txt
the $shared_loci_file can be grenerated through novelty/identify_overlap_loci.sh
```
