## Contents

* [mixer_uni_table.sh](#mixer_uni_tablesh)
* [mixer_bi_table.sh](#mixer_bi_tablesh)
* [fdr_fuma_input.sh](#fdr_fuma_inputsh)
* [fdr_fuma_snp_table.sh](#fdr_fuma_snp_tablesh)
* [fdr_fuma_loci_table.sh](#fdr_fuma_loci_tablesh)
* [fuma_gene_table.sh](#fuma_gene_tablesh)

## Summary of scripts

### mixer_uni_table.sh

**Function**
This script extracts data from mixer analysis for supplementary univariate
table.

**Usage** ``sh mixer_uni_table.sh uni_csv outfile``

**Arguments**
* `uni_csv` - csv file containing univariate info
* `outfile` - output file

### mixer_bi_table.sh

**Function**
This script extracts data from mixer analysis for supplementary bivariate
table.

**Usage** ``sh mixer_bi_table.sh bi_csv outfile``

**Arguments**
* `bi_csv` - csv file containing bivariate info
* `outfile` - output file

### fdr_fuma_input.sh

**Function**
This script generates a snp list for FUMA analysis.

**Usage** ``sh fdr_fuma_input.sh fdr_clump_snp_file fdr r2 outfile``

**Arguments**
* `fdr_clump_snp_file` - fdr clumping snp file
* `fdr` - FDR filter for selecting snps
* `r2` - r2 filter for selecting snps
* `outfile` - output file including selected snp list

**Example**
```
sh fdr_fuma_input.sh conj.result.clump.snps.csv 0.1 0.6 trait1_vs_trait2_input_snps.txt
```

### fdr_fuma_snp_table.sh

**Function**
This script generates data for supplementary snp table (FUMA+SUMSTAT)
based on selected snp subset (fdr,r2) from FDR/FUMA analysis.
NOTE: there might be difference between the loci numbers from FDR analysis
and the ones from FUMA analysis (GenomicLocus). This is because the former
is based on clumping of FDR, while the latter is based on clumping of
p-value, good to distiguish between them.

**Usage** ``sh fdr_fuma_snp_table.sh fdr_clump_snp_file fdr r2 fuma_snp_file sumstat1 sumstat2 tag1 tag2 outfolder``

**Arguments**
* `fdr_clump_snp_file` - fdr clumping snp file
* `fdr` - FDR filter for selecting snps
* `r2` - r2 filter for selecting snps
* `fuma_snp_file` - fuma annotated snp file
* `sumstat1` - primary summary statistic file
* `sumstat2` - secondary summary statistic file
* `tag1` - tag of trait1
* `tag2` - tag of trait2
* `outfolder` - output folder

**Example**
```
sh fdr_fuma_snp_table.sh conj.result.clump.snps.csv 0.1 0.6 snps.txt sumstat/std/trait1.sumstats.gz sumstat/std/trait2.sumstats.gz tag1 tag2 outfolder
```

### fdr_fuma_loci_table.sh

**Function**
This script generates data for supplementary loci table (FDR+FUMA+SUMSTAT)
from FDR/FUMA analysis.

**Usage** ``sh fdr_fuma_loci_table.sh fdr_clump_loci_file fdr_fuma_snp_table outfile [trait1locifile trait2locifile] [fdr_snp_table fuma_gwasc keyword1 keyword2]``

**Arguments**
* `fdr_clump_loci_file` - fdr clumping loci file
* `fdr_fuma_snp_table` - trait1_vs_trait2_snps.txt file generated by script fdr_fuma_snp_table.sh
* `outfile` - output file
* `trait1locifile` - old loci file with 3 or 4 columns (CHR, MinBP, MaxBP, [leadSNP]) of trait1, use '-' for none
* `trait2locifile` - old loci file with 3 or 4 columns (CHR, MinBP, MaxBP, [leadSNP]) of trait2, use '-' for none
* `fdr_snp_table` - fdr clumping snp file, use '-' for none
* `fuma_gwasc` - fuma gwas catalog file, use '-' for none
* `keyword1` - keyword of trait1, use '-' for none
* `keyword2` - keyword of trait2, use '-' for none

**Example**
```
sh fdr_fuma_loci_table.sh conj.result.clump.loci.csv trait1_vs_trait2_snps_conj.txt trait1_vs_trait2_loci.txt trait1_old_loci.csv trait2_old_loci.csv conj.result.clump.snps.csv gwascatalog.txt keyword1 keyword2
sh fdr_fuma_loci_table.sh conj.result.clump.loci.csv adhd_vs_mood_snps_conj.txt adhd_vs_mood_loci.txt adhd_gwas_loci.csv mood_gwas_loci.csv conj.result.clump.snps.csv gwascatalog.txt 'attention deficit' 'mood instability'
```

### fuma_gene_table.sh

**Function**
This script generates data for supplementary gene table from FUMA analysis.

**Usage** ``sh fuma_gene_table.sh fuma_gene_file outfile``

**Arguments**
* `fuma_gene_file` - file that contains fuma gene info
* `outfile` - output file

**Example**
```
sh fuma_gene_table.sh genes.txt trait1_vs_trait2_genes.txt
```

### fdr_fuma_eqtl_table.sh

**Function**
This script generates data for supplementary eqtl table from FDR/FUMA analysis.

**Usage** ``sh fdr_fuma_eqtl_table.sh fdr_clump_loci_file fuma_eqtl_file outfile``

**Arguments**
* `fdr_clump_loci_file` - fdr clumping loci file
* `fuma_eqtl_table` - file that contains fuma eqtl info
* `outfile` - output file

**Example**
```
sh fuma_eqtl_table.sh conj.result.clump.loci.csv eqtl.txt trait1_vs_trait2_eqtl.txt
```
